sync2/poller.go

```diff
@@ -791,8 +793,47 @@ func (p *poller) parseRoomsResponse(ctx context.Context, res *SyncResponse) erro
 			stateCalls++
 			prependStateEvents, err := p.receiver.Initialise(ctx, roomID, roomData.State.Events)
 			if err != nil {
-				lastErrs = append(lastErrs, fmt.Errorf("Initialise[%s]: %w", roomID, err))
-				continue
+				_, ok := err.(*internal.DataError)
+				if ok {
+					// This typically happens when we are missing an m.room.create event.
+					// Synapse may sometimes send the m.room.create event erroneously in the timeline,
+					// so check if that is the case here. See https://github.com/matrix-org/complement/pull/690
+					var createEvent json.RawMessage
+					for i, ev := range roomData.Timeline.Events {
+						if gjson.ParseBytes(ev).Get("type").Str == "m.room.create" {
+							createEvent = roomData.Timeline.Events[i]
+							// remove the create event from the timeline so we don't double process it
+							roomData.Timeline.Events = append(roomData.Timeline.Events[:i], roomData.Timeline.Events[i+1:]...)
+							break
+						}
+					}
+					if createEvent != nil {
+						roomData.State.Events = slices.Insert(roomData.State.Events, 0, createEvent)
+						// retry the processing of the room state
+						prependStateEvents, err = p.receiver.Initialise(ctx, roomID, roomData.State.Events)
+						if err == nil {
+							const warnMsg = "parseRoomsResponse: m.room.create event was found in the timeline not state"
+							logger.Warn().Str("user_id", p.userID).Str("room_id", roomID).Int(
+								"timeline", len(roomData.Timeline.Events),
+							).Int("state", len(roomData.State.Events)).Msg(warnMsg)
```

Comment:
Might want to note that these numbers are after we moved the create event into the state block.